src/odm_sharing/private/trees.py

```diff
@@ -0,0 +1,457 @@
+        op = parse_op(ctx, op_str)
+        is_range = (op == Op.RANGE)
+        values = \
+            parse_list(ctx, val_str, min=2, max=2) if is_range else [val_str]
+        key_node = Node(rule_id=rule_id, kind=NodeKind.KEY, value=key)
+        value_nodes = to_value_nodes(rule_id, values)
+        return Node(
+            rule_id=rule_id,
+            kind=NodeKind.FILTER,
+            value=op_str,
+            sons=([key_node] + value_nodes),
+        )
+    elif mode == RuleMode.GROUP:
+
+        def not_filter_group(node: Node) -> bool:
+            return node.kind not in [NodeKind.FILTER, NodeKind.GROUP]
+
+        op = parse_op(ctx, op_str)
+        if op not in [Op.AND, Op.OR]:
+            fail(ctx, 'incompatible group operator')
+        ids = parse_int_list(ctx, val_str, min=2)
+        sons = seq(ids).map(get_ctx_node)
+        if sons.map(not_filter_group).any():
+            fail(ctx, 'group-rules can only refer to other filter/group-rules')
+        return Node(
+            rule_id=rule_id,
+            kind=NodeKind.GROUP,
+            value=op.value,
+            sons=sons,
+        )
+    elif mode == RuleMode.SHARE:
+
+        def is_select(node: Node) -> bool:
+            return node.kind == NodeKind.SELECT
+
+        def init_table_node(share_value_nodes: List[Node],
+                            table: str, select_id: int) -> Node:
+            assert select_id
+            select_node = get_ctx_node(select_id)
+            filter_root = get_filter_root(ctx, table, share_value_nodes)
+            return Node(
+                rule_id=select_id,
+                kind=NodeKind.TABLE,
+                value=table,
+                sons=([select_node] + ([filter_root] if filter_root else [])),
+            )
+
+        def init_share_node(rule_id: int, table_nodes: List[Node], org: str
+                            ) -> Node:
+            return Node(
+                rule_id=rule_id,
+                kind=NodeKind.SHARE,
+                value=org,
+                sons=table_nodes,
+            )
+
+        orgs = parse_list(ctx, key, min=1)
+        ids = parse_int_list(ctx, val_str, min=1)
+        share_value_nodes = seq(ids).map(get_ctx_node).list()
+        select_nodes = seq(share_value_nodes).filter(is_select).list()
+        table_select_ids = get_table_select_ids(ctx, select_nodes)
+
+        init_table_node2 = partial(init_table_node, share_value_nodes)
+        table_nodes = seq(table_select_ids.items())\
+            .smap(init_table_node2)\
+            .list()
+
+        init_share_node2 = partial(init_share_node, rule_id, table_nodes)
+        share_nodes = seq(orgs).map(init_share_node2).list()
+
+        root_node_candidate = Node(
+            rule_id=0,
+            kind=NodeKind.ROOT,
+            sons=share_nodes
+        )
+        return root_node_candidate
+    else:
+        assert False, 'not all cases covered'
+
+
+def add_node(ctx: Ctx, rule_id: RuleId, table: str, mode: RuleMode,
```

Comment:
Can we remove the `rule_id`, `mode`, `key`, `op_str`, and `val_str` parameters in favor of passing in the Rule object?

Reply:
then I'd have to change the code structure since this function is meant to only add a single node, and a rule can produce multiple nodes if it has multiple tables. I give all the parameters individually since the table param isn't equal to `rule.table`. It would become `add_nodes_from_rule` and the table iteration would be moved in there, making it a more complex function. What do you think?

Reply:
That logic makes sense, its just a lot of parameters haha.